resteasy-core/src/main/java/org/jboss/resteasy/plugins/server/servlet/HttpServletResponseWrapper.java

```diff
@@ -56,13 +164,111 @@ public void close() throws IOException
       {
          //NOOP (RESTEASY-1650)
       }
+
+      @Override
+      public CompletionStage<Void> asyncFlush()
+      {
+         AsyncOperation op = new FlushOperation(this);
+         queue(op);
+         return op.future;
+      }
+
+      @Override
+      public CompletionStage<Void> asyncWrite(byte[] bytes, int offset, int length)
+      {
+         AsyncOperation op = new WriteOperation(this, bytes, offset, length);
+         queue(op);
+         return op.future;
+      }
+
+      private void queue(AsyncOperation op)
+      {
+         // fetch it from the context directly to avoid having to restore the context just in case we're invoked on a context-less thread
+         HttpRequest resteasyRequest = (HttpRequest) contextDataMap.get(HttpRequest.class);
+         if(request.isAsyncStarted() && !resteasyRequest.getAsyncContext().isOnInitialRequest()) {
+            synchronized(this) {
+               ServletOutputStream os;
+               try
+               {
+                  os = response.getOutputStream();
+               } catch (IOException e)
+               {
+                  // return a failed future, do not queue it
+                  op.future.completeExceptionally(e);
+                  return;
+               }
+               if(!asyncRegistered) {
+                  // start the queue
+                  asyncRegistered = true;
+                  // make sure we have something ready to be executed
+                  addToQueue(op);
+                  os.setWriteListener(this);
+               } else if(os.isReady()) {
+                  // it's possible that we startAsync and queue, then queue another event and the stream becomes ready before
+                  // onWritePossible is called, which means we need to flush the queue here to guarantee ordering if that happens
+                  flushQueue(os);
+                  // we're not allowed to queue work if the output is ready
+                  lastAsyncOperation = op;
+                  op.work(os);
+               } else {
+                  // just queue
+                  addToQueue(op);
+               }
+            }
+         } else {
+            op.work(null);
```

Comment:
How is this ok?  We seem to be getting here and no bytes are happening and the thread hangs waiting for a future that will never complete to happen.  It does appear the bytes are here.

Reply:
@crankydillo What version are you using? I made some changes around here recently. But it should be completed here https://github.com/resteasy/resteasy/pull/2170/files/3c11dd4f389837dc22474e6b4f4d23926a88ccea#diff-a2be1ee2e763d57c738434047303d73e02cf6171a606b8dc34cf3fcc0aa3b3d6R104-R106